File: BrazeProject/ios/BrazeDynamicConfigurationBridge/BrazeDynamicConfigurationBridge.m
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_REMAP_MODULE(BrazeDynamicConfigurationBridge, BrazeDynamicConfigurationBridge, NSObject)
RCT_EXTERN_METHOD(saveConfig: (NSDictionary)configOptions resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(initializeWithSavedConfig: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
@end

File: BrazeProject/ios/BrazeDynamicConfigurationBridge/BrazeDynamicConfigurationBridge.swift
import Foundation

@objc(BrazeDynamicConfigurationBridge)
public class BrazeDynamicConfigurationBridge: NSObject {
  @objc static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  @objc func saveConfig(
    _ configOptions: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      try BrazeDynamicConfiguration.saveConfig(fromDictionary: configOptions)
      
      resolve(nil)
    } catch let error as NSError {
      reject(error.domain, error.localizedDescription, error)
    } catch let error {
      reject("BrazeDynamicConfigurationBridge", "Saving failed", error)
    }
  }
  
  @objc func initializeWithSavedConfig(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        try await BrazeDynamicConfiguration.initializeWithSavedConfig()
        
        resolve(nil)
      } catch let error as NSError {
        reject(error.domain, error.localizedDescription, error)
      } catch let error {
        reject("BrazeDynamicConfigurationBridge", "Saving failed", error)
      }
    }
  }
}
